### PR TITLE
react: Make ValidationMap only accept keys from T.

### PR DESCRIPTION
**Pull request by seansfkelley**
Please fill in this template.

- [x] Make your PR against the `master` branch.
- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `tsc` without errors.
- [x] Run `npm run lint package-name` if a `tslint.json` is present.

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://facebook.github.io/react/docs/typechecking-with-proptypes.html
- [x] Increase the version number in the header if appropriate.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "../tslint.json" }`.

### CODE REVIEW

**Comment by vsiao** [comment]
This is probably fine, but I'm wondering why you need it? In my experience the type parameter `T` of `ValidationMap` isn't very useful because the whole point of a validation map is to make runtime checks on the type of `T`.

**Comment by seansfkelley** [comment]
The idea was that you don't want to validate things that you aren't typed to receive. I hit an issue where I renamed a prop but non-typescript consumers weren't getting updated validations because the typings were too lenient and didn't force me to rename it. 

I don't know if there would be a use-case for allowing validation of props you haven't declared.

**Comment by paulvanbrenk** [comment]
Thanks for the PR, it looks like there were some issues in the CI. Can you take a look at those:

```
=== ERRORS ===



Error in react-autosuggest
Command failed: node /home/travis/build/DefinitelyTyped/DefinitelyTyped/node_modules/types-publisher/node_modules/tslint/lib/index.js/../tslint-cli.js --format stylish index.d.ts react-autosuggest-tests.tsx

react-autosuggest-tests.tsx
136:89   whitespace  missing whitespace
148:71   whitespace  missing whitespace
319:69   whitespace  missing whitespace
418:113  whitespace  missing whitespace



Error in react-bootstrap-daterangepicker
Command failed: node /home/travis/build/DefinitelyTyped/DefinitelyTyped/node_modules/types-publisher/bin/tester/../../typescript-installs/2.1/node_modules/typescript/lib/tsc.js 

../daterangepicker/index.d.ts(7,25): error TS7016: Could not find a declaration file for module 'moment'. '/home/travis/build/DefinitelyTyped/DefinitelyTyped/node_modules/moment/moment.js' implicitly has an 'any' type.



Error in react-daterange-picker
Command failed: node /home/travis/build/DefinitelyTyped/DefinitelyTyped/node_modules/types-publisher/bin/tester/../../typescript-installs/2.1/node_modules/typescript/lib/tsc.js 

../moment-range/index.d.ts(6,25): error TS7016: Could not find a declaration file for module 'moment'. '/home/travis/build/DefinitelyTyped/DefinitelyTyped/node_modules/moment/moment.js' implicitly has an 'any' type.
../moment-range/index.d.ts(10,16): error TS2665: Invalid module name in augmentation. Module 'moment' resolves to an untyped module at '/home/travis/build/DefinitelyTyped/DefinitelyTyped/node_modules/moment/moment.js', which cannot be augmented.
index.d.ts(8,25): error TS2497: Module '"/home/travis/build/DefinitelyTyped/DefinitelyTyped/moment-range/index"' resolves to a non-module entity and cannot be imported using this construct.
react-daterange-picker-tests.tsx(4,25): error TS2497: Module '"/home/travis/build/DefinitelyTyped/DefinitelyTyped/moment-range/index"' resolves to a non-module entity and cannot be imported using this construct.
```

**Comment by seansfkelley** [comment]
I didn't touch any of those files, but I rebased my commit on top of the latest commit on master as of just now with the assumption that they were broken earlier (or perhaps build caching?).

**Comment by andy-ms** [comment]
The process is to fix the errors, of course! #15288 fixes the lint errors. Microsoft/types-publisher#322 should fix the other errors. Please rebase onto the latest master.

File: react/index.d.ts [hunk 1]
```diff
@@ -2610,9 +2610,7 @@ declare namespace React {
         isRequired: Validator<T>;
     }
 
-    interface ValidationMap<T> {
-        [key: string]: Validator<T>;
-    }
+    type ValidationMap<T> = { [K in keyof T]?: Validator<T> };
```
Comment (andy-ms) [inline react/index.d.ts]: Is the value type really `Validator<T>` and now `Validator<T[K]>`? Not familiar with the library but the latter is more common.

Comment (seansfkelley) [inline react/index.d.ts]: Yes; the validator functions receive all of `T` along with a `string` parameter that specified which key that function is being asked to validate. Good question though!

**Comment by seansfkelley** [comment]
@andy-ms looks like I had the right idea, but the wrong timing. Thanks for fixing the linting errors, etc! Rebased onto master again just now.

**Comment by paulvanbrenk** [comment]
thanks @andy-ms !

**Pull request merged**